// sdmx-rest/lib/metadata/metadata-type.js
// Generated by CoffeeScript 1.10.0
(function() {
  var itemSchemes, types,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  itemSchemes = ['categoryscheme', 'conceptscheme', 'codelist', 'organisationscheme', 'agencyscheme', 'dataproviderscheme', 'dataconsumerscheme', 'organisationunitscheme'];

  types = {
    DATA_STRUCTURE: 'datastructure',
    METADATA_STRUCTURE: 'metadatastructure',
    CATEGORY_SCHEME: 'categoryscheme',
    CONCEPT_SCHEME: 'conceptscheme',
    CODELIST: 'codelist',
    HIERARCHICAL_CODELIST: 'hierarchicalcodelist',
    ORGANISATION_SCHEME: 'organisationscheme',
    AGENCY_SCHEME: 'agencyscheme',
    DATA_PROVIDER_SCHEME: 'dataproviderscheme',
    DATA_CONSUMER_SCHEME: 'dataconsumerscheme',
    ORGANISATION_UNIT_SCHEME: 'organisationunitscheme',
    DATAFLOW: 'dataflow',
    METADATAFLOW: 'metadataflow',
    REPORTING_TAXONOMY: 'reportingtaxonomy',
    PROVISION_AGREEMENT: 'provisionagreement',
    STRUCTURE_SET: 'structureset',
    PROCESS: 'process',
    CATEGORISATION: 'categorisation',
    CONTENT_CONSTRAINT: 'contentconstraint',
    ATTACHMENT_CONSTRAINT: 'attachmentconstraint',
    STRUCTURE: 'structure'
  };

  exports.MetadataType = Object.freeze(types);

  exports.isItemScheme = function(type) {
    return indexOf.call(itemSchemes, type) >= 0;
  };

}).call(this);
