// sdmx-rest/lib/service/service.js
// Generated by CoffeeScript 1.10.0
(function() {
  var ApiVersion, DataFormat, Service, createErrorMessage, createSecureInstance, defaults, isValidEnum, isValidService, isValidUrl, ref, service;

  ApiVersion = require('../utils/api-version').ApiVersion;

  ref = require('../utils/validators'), isValidEnum = ref.isValidEnum, createErrorMessage = ref.createErrorMessage;

  DataFormat = require('../data/data-format').DataFormat;

  defaults = {
    api: ApiVersion.LATEST
  };

  isValidUrl = function(url, errors) {
    var valid;
    valid = url;
    if (!valid) {
      errors.push(url + " is not in a valid url");
    }
    return valid;
  };

  isValidService = function(q) {
    var errors, isValid;
    errors = [];
    isValid = isValidUrl(q.url, errors) && isValidEnum(q.api, ApiVersion, 'versions of the SDMX RESTful API', errors);
    return {
      isValid: isValid,
      errors: errors
    };
  };

  createSecureInstance = function(service) {
    var key, secure;
    secure = {};
    for (key in service) {
      secure[key] = service[key];
    }
    secure.url = secure.url.replace('http', 'https');
    return secure;
  };

  service = Service = (function() {
    function Service() {}

    Service.ECB = {
      id: 'ECB',
      name: 'European Central Bank',
      api: ApiVersion.v1_0_2,
      url: 'http://sdw-wsrest.ecb.europa.eu/service',
      format: DataFormat.SDMX_JSON
    };

    Service.SDMXGR = {
      id: 'SDMXGR',
      name: 'SDMX Global Registry',
      api: ApiVersion.v1_1_0,
      url: 'http://registry.sdmx.org/FusionRegistry/ws/rest'
    };

    Service.EUROSTAT = {
      id: 'EUROSTAT',
      name: 'Eurostat',
      api: ApiVersion.v1_0_2,
      url: 'http://www.ec.europa.eu/eurostat/SDMX/diss-web/rest'
    };

    Service.OECD = {
      id: 'OECD',
      name: 'Organisation for Economic Co-operation and Development',
      api: ApiVersion.v1_0_2,
      url: 'http://stats.oecd.org/SDMX-JSON'
    };

    Service.WB = {
      id: 'WB',
      name: 'World Bank',
      api: ApiVersion.v1_0_2,
      url: 'http://wits.worldbank.org/API/V1/SDMX/V21/rest'
    };

    Service.ECB_S = createSecureInstance(Service.ECB);

    Service.SDMXGR_S = createSecureInstance(Service.SDMXGR);

    Service.OECD_S = createSecureInstance(Service.OECD);

    Service.from = function(opts) {
      var input, ref1;
      service = {
        id: opts != null ? opts.id : void 0,
        name: opts != null ? opts.name : void 0,
        url: opts != null ? opts.url : void 0,
        api: (ref1 = opts != null ? opts.api : void 0) != null ? ref1 : defaults.api,
        format: opts != null ? opts.format : void 0
      };
      input = isValidService(service);
      if (!input.isValid) {
        throw Error(createErrorMessage(input.errors, 'service'));
      }
      return service;
    };

    return Service;

  })();

  exports.Service = service;

}).call(this);
