// sdmx-rest/lib/utils/sdmx-patterns.js
// Generated by CoffeeScript 1.10.0
(function() {
  var FlowRefType, IDType, IDTypeAlone, NCNameIDType, NCNameIDTypeAlone, NestedIDType, NestedIDTypeAlone, NestedNCNameIDType, NestedNCNameIDTypeAlone, ProviderRefType, ReportingPeriodType, SeriesKeyType, VersionNumber, VersionType, VersionTypeAlone;

  NCNameIDType = /[A-Za-z][A-Za-z0-9_\-]*/;

  NCNameIDTypeAlone = RegExp("^" + NCNameIDType.source + "$");

  NestedNCNameIDType = RegExp(NCNameIDType.source + "(\\." + NCNameIDType.source + ")*");

  NestedNCNameIDTypeAlone = RegExp("^" + NestedNCNameIDType.source + "$");

  IDType = /[A-Za-z0-9_@$\-]+/;

  IDTypeAlone = RegExp("^" + IDType.source + "$");

  VersionNumber = /[0-9]+(\.[0-9]+)*/;

  VersionType = RegExp("(all|latest|" + VersionNumber.source + ")");

  VersionTypeAlone = RegExp("^" + VersionType.source + "$");

  NestedIDType = /[A-Za-z0-9_@$\-]+(\.[A-Za-z0-9_@$\-]+)*/;

  NestedIDTypeAlone = RegExp("^" + NestedIDType.source + "$");

  SeriesKeyType = RegExp("^(" + IDType.source + "([+]" + IDType.source + ")*)?([.](" + IDType.source + "([+]" + IDType.source + ")*)?)*$");

  FlowRefType = RegExp("^(" + IDType.source + "|(" + NestedNCNameIDType.source + "\\," + IDType.source + "(\\,(latest|(" + VersionNumber.source + ")))?))$");

  ProviderRefType = RegExp("^(" + NestedNCNameIDType.source + ",)?" + IDType.source + "$");

  ReportingPeriodType = /^\d{4}\-([ASTQ]\d{1}|[MW]\d{2}|[D]\d{3})$/;

  exports.NCNameIDType = NCNameIDTypeAlone;

  exports.NestedNCNameIDType = NestedNCNameIDTypeAlone;

  exports.IDType = IDTypeAlone;

  exports.VersionType = VersionTypeAlone;

  exports.NestedIDType = NestedIDTypeAlone;

  exports.FlowRefType = FlowRefType;

  exports.ProviderRefType = ProviderRefType;

  exports.ReportingPeriodType = ReportingPeriodType;

  exports.SeriesKeyType = SeriesKeyType;

}).call(this);
